nixos/modules/config/system-path.nix

```diff
@@ -6,50 +6,209 @@
+            if length values == 0 then
+              abort "Something went terribly wrong"
+            else {
+              # If we get multiple values here, it means that we got the same
+              # derivation multiple times.
+              # This is usually because different outputs were selected.
+              # Therefore we add the derivation once and collect all the selected outputs.
+              package = defaultOutput (head values);
+              outputs = lib.foldl (acc: drv: acc // outputsFor' drv) {} values;
+            }
+          )
+          (lib.groupBy (drv: builtins.unsafeDiscardStringContext drv.drvPath) list);
+    in
+    types.coercedTo
+      (types.listOf types.package)
+      pkgsToAttrs
+      (types.attrsOf (types.coercedTo
+        types.package
+        drvToAttrs
+        (types.submodule ( { config, lib, ... }: {
+          options = {
+            enable = (lib.mkEnableOption "the package") // {
+              default = true;
+            };
+            package = lib.mkOption {
+              # We can get the same derivation here if it was added with different
+              # outputs selected. In that case we want to merge the values if the
+              # outpath is the same, and concat the lists of outputs.
+              type = types.packageByOutPath;
+              description = "The derivation that will be used for this package.";
+            };
+            outputs =
+              let
+                # config gets shadowed below, so we assign it another name here
+                # to be able to access it still.
+                outerConfig = config;
+              in
+              lib.mkOption {
+                type = types.attrsOf (types.submodule ( { name, config, lib, ... }: {
+                  options = {
+                    enable = lib.mkEnableOption "the output" // {
+                      default = true;
+                    };
+                    name = lib.mkOption {
+                      type = types.str;
+                      default = name;
+                      description = "The name of this output.";
+                    };
+                    priority = lib.mkOption {
+                      type = types.nullOr (types.int // {
+                        # If an output gets selected twice with different priorities,
+                        # we want to retain the higher priority.
+                        merge = loc: defs:
+                          if defs == [] then abort "This case should never happen."
+                          else if length defs == 1 then (head defs).value
+                          else
+                            lib.foldl' (sum: el: lib.max sum el.value) (head defs).value (tail defs);
+                      });
+                      default = outerConfig.package.${name}.meta.priority or null;
+                      defaultText = ''
+                        The meta.priority value of the package, or null if there is no priority assigned.
```

Comment:
Could you describe the effect of `priority`?
E.g. what does a large value do? Do packages without a priority have a default priority?
I think you could link to https://nixos.org/manual/nixpkgs/stable/#var-meta-priority, but that doc isn't much better.
Maybe this? https://nixos.org/manual/nix/stable/command-ref/nix-env/install.html?highlight=priority#description

I think it'd be ok to copy some of this information, because
 - NixOS/nixpkgs duplicates this logic anyway
 - It's not going to change in Nix, because that'd break profiles for no good reason